File: modules/uml/app/model/UmlExerciseReader.java
package model;

import com.fasterxml.jackson.databind.JsonNode;

import model.exercisereading.ExerciseReader;
import play.data.DynamicForm;

public class UmlExerciseReader extends ExerciseReader<UmlExercise> {

  private static final UmlExerciseReader INSTANCE = new UmlExerciseReader();

  private UmlExerciseReader() {
    super("uml", UmlExercise.finder, UmlExercise[].class);
  }

  public static UmlExerciseReader getInstance() {
    return INSTANCE;
  }

  @Override
  public void initRemainingExFromForm(UmlExercise exercise, DynamicForm form) {
    exercise.setClassSelText(form.get("classSelText"));
    exercise.setDiagDrawText(form.get("diagDrawText"));
    exercise.setSolution(form.get("solution"));
  }

  @Override
  public UmlExercise instantiateExercise(int id) {
    return new UmlExercise(id);
  }

  @Override
  protected void updateExercise(UmlExercise exercise, JsonNode exerciseNode) {
    final String text = "";

    exercise.setMappings(readArray(exerciseNode.get(StringConsts.MAPPINGS_NAME), Mapping::fromJson));
    exercise.setToIgnore(readTextArray(exerciseNode.get(StringConsts.IGNORE_NAME), "#"));

    final UmlExTextParser parser = new UmlExTextParser(text, exercise.getMappings(), exercise.getToIgnore());
    exercise.setClassSelText(parser.parseTextForClassSel());
    exercise.setDiagDrawText(parser.parseTextForDiagDrawing());

    // Save solution as json in db
    exercise.setSolution(exerciseNode.get(StringConsts.SOLUTION_NAME).asText());
  }

}

File: modules/uml/app/model/UmlExerciseReader.scala
package model

import com.fasterxml.jackson.databind.JsonNode
import model.StringConsts.{IGNORE_WORDS_NAME, MAPPINGS_NAME, SOLUTION_NAME}
import model.exercisereading.{ExerciseReader, JsonReader}

import scala.collection.JavaConverters._


object UmlExerciseReader extends ExerciseReader[UmlExercise]("uml", UmlExercise.finder, classOf[Array[UmlExercise]]) {

//  override def initRemainingExFromForm(exercise: UmlExercise, form: DynamicForm) {
  //    exercise.classSelText = form.get("classSelText")
  //    exercise.diagDrawText = form.get("diagDrawText")
  //    exercise.solution = form.get("solution")
  //  }

  override def instantiate(id: Int): UmlExercise = {
    val ex = new UmlExercise()
    ex.id = id
    ex
  }

  override def updateExercise(exercise: UmlExercise, node: JsonNode) {
    val mappings = JsonReader.readMap(node.get(MAPPINGS_NAME))
    val ignoreWords = JsonReader.readTextArray(node.get(IGNORE_WORDS_NAME))

    exercise.mappings = mappings.asJava
    exercise.ignoreWords = ignoreWords.asJava


    val parser = new UmlExTextParser(exercise.text, mappings, ignoreWords)
    exercise.classSelText = parser.parseTextForClassSel
    exercise.diagDrawText = parser.parseTextForDiagDrawing

    // Save solution as json in db
    exercise.solution = node.get(SOLUTION_NAME).toString
  }

}